### ISSUE TITLE
Issue with status bar on iPhone X when toolbar is at the top (v3.0.0)

### ISSUE THREAD

**Issue by russcarver** [apache/cordova-plugin-inappbrowser#301]
There is an issue with the status bar and the iPhone X as described at http://www.doroja.com/smartdic.aspx?decode=1&url=aHR0cHM6Ly9zdGFja292ZXJmbG93LmNvbS9xdWVzdGlvbnMvNDczOTk5MzgvaXNzdWUtd2l0aC1zdGF0dXMtYmFyLXVzaW5nLWNvcmRvdmEtaW5hcHBicm93c2VyLWFuZC1pb3MtMTEtYW5kLWlwaG9uZS14B64Coded!

When the inappbrowser toolbar is at the top of the screen, the status bar on the iPhone X is half white and half gray.

Changing the following line in CDVInAppBrowser.m solves the problem and makes the whole status bar gray.  Change line:
`statusBarFrame.size.height = STATUSBAR_HEIGHT;`
to
`statusBarFrame.size.height = 0;`

This does not seem to negatively affect other iPhones or iPads.

**Comment by russcarver** [comment]
<img width="391" alt="screen shot 2018-09-21 at 8 12 49 pm" src="https://user-images.githubusercontent.com/4061972/45912721-e3f7fb00-bdda-11e8-9487-a83800e5940b.png">

**Comment by russcarver** [comment]
<img width="407" alt="screen shot 2018-09-21 at 7 39 01 pm" src="https://user-images.githubusercontent.com/4061972/45912729-eb1f0900-bdda-11e8-810a-e6591080587c.png">

**Comment by vespino** [comment]
@russcarver is this something I can fix when using Phonegap Build? Or is this an update waiting to happen?

**Comment by vespino** [comment]
@russcarver were you able to solve this issue? Adding the splashscreens does not solve it and iPhone X has been around for over a year now, so you would think the guys at Phonegap would have fixed this by now.

**Comment by russcarver** [comment]
@vespino I'm waiting for them to fix this. In the mean time, I've just forked it and made the changes shown above.

**Comment by vespino** [comment]
@russcarver can I use your fork?

**Comment by russcarver** [comment]
@vespino I don't have it anymore as I decided to go a different route, but if you fork it yourself and make the change outlined above, it will work fine. You can install plugins from local directories using the "cordova plugin add" command.

**Comment by vespino** [comment]
@russcarver do you mind sharing the different route btw?

**Comment by vespino** [comment]
@russcarver think I have found it in this file: https://github.com/vespino/cordova-plugin-inappbrowser/blob/master/src/ios/CDVInAppBrowserNavigationController.m

**Comment by russcarver** [comment]
Mine was in the other file, but it could be outdated by now.   You can also just change the #define of the constant since it looks used in only that one place.

**Comment by vespino** [comment]
@russcarver seems like the change I made indeed solves the statusbar issue, but raises another one. When opening a link in the inappbrowser and closing it, the app freezes on iOS. You would think the guys at phonegap would have addressed this issue more than a year after introduction of the iPhone X.

**Comment by russcarver** [comment]
@vespino Sorry to hear that. That didn't happen to me in my tests, but I didn't test it out on iOS 12.  Any details about your environment shared here may help the code owner debug.

**Comment by vespino** [comment]
@russcarver I'm using this fork of themeable browser now: https://github.com/toontoet/cordova-plugin-themeablebrowser.git

**Comment by mtlehtin** [comment]
Besides this fix just pulls the transparent grayish backdrop/overlay to top of the viewport and not the inappbrowser-view.

But how to pull the entire inappbrowser-view under the notch? Is that even possible?

**Comment by mtlehtin** [comment]
In addition to setting `statusBarFrame.size.height = 0;` i also ended up doing something like following that does not entirely rely on the hardcoded STATUSBAR_HEIGHT. Following code adds top and bottom safeareas for the view (Since we have no control for 3rd party app that we are using inside IAB and it is easier to handle in the InAppBrowser).

CDVWKInAppBrowser.m:
```
- (BOOL)hasTopNotch {
    if (@available(iOS 11.0, *)) {
        return [[[UIApplication sharedApplication] delegate] window].safeAreaInsets.top > 20.0;
    }

    return  NO;
}

- (void)viewWillAppear:(BOOL)animated
{
    if (IsAtLeastiOSVersion(@"7.0") && !viewRenderedAtLeastOnce) {
        viewRenderedAtLeastOnce = TRUE;
        CGRect viewBounds = [self.webView bounds];
        
        if ([self hasTopNotch]) {
            BOOL toolbarVisible = !self.toolbar.hidden;
            BOOL toolbarIsAtBottom = ![_browserOptions.toolbarposition isEqualToString:kInAppBrowserToolbarBarPositionTop];

            float topSafeArea = [[[UIApplication sharedApplication] delegate] window].safeAreaInsets.top;
            float bottomSafeArea = [[[UIApplication sharedApplication] delegate] window].safeAreaInsets.bottom;

            if (toolbarVisible && toolbarIsAtBottom) {
                bottomSafeArea = 0.0;
            }

            viewBounds.origin.y = topSafeArea;
            viewBounds.size.height = viewBounds.size.height - (topSafeArea + bottomSafeArea);
        } else {
            viewBounds.origin.y = STATUSBAR_HEIGHT;
            viewBounds.size.height = viewBounds.size.height - STATUSBAR_HEIGHT;
        }
        self.webView.frame = viewBounds;
        [[UIApplication sharedApplication] setStatusBarStyle:[self preferredStatusBarStyle]];
    }
    [self rePositionViews];
    
    [super viewWillAppear:animated];
}
```

The logic for hasTopNotch is from here: https://stackoverflow.com/questions/46192280/detect-if-the-device-is-iphone-x.

In addition to that, I removed they gray-background: `self.view.backgroundColor = [UIColor grayColor];` and replaced it with configurable option for background color and UIStatusBarStyleLightContent or UIStatusBarStyleDefault.

**Comment by KaiCMueller** [comment]
@mtlehtin I would also appreciate a pull request for your solution

**Comment by GreenSpecialist** [comment]
inappbrowser-wkwebview is deprecated and it is merged with inappbrowser. There is no CDVWKInAppBrowser.m file anymore and if you implement @mtlehtin change in the CDVInAppBrowser.m it doesnt fix the issue... Any help?

**Comment by mosababubakr** [comment]
@dpa99c please take a look to this

**Comment by NiklasMerz** [comment]
We solved this that way: https://github.com/GEDYSIntraWare/cordova-plugin-inappbrowser/commit/0c75a939327bd2598f6dc85644df60f9835e5ac3

Should I create a PR?

**Comment by mosababubakr** [comment]
Hello @NiklasMerz 

If you test your solution and work perfect on iphone x and previous version of iphone, you can create a PR for this solution, because this issue still not solved for iphone x family devices.

Regards

**Comment by NiklasMerz** [comment]
Thanks @mosababubakr for testing. PR created. Thanks @mtlehtin for your code, I adapted if a little bit.

Looking forward for reviews.

**Comment by mosababubakr** [comment]
@NiklasMerz  please ask @dpa99c  to test your solution (WKWEBVIEW mode) and put your PR to be approved.

Regards

**Comment by NiklasMerz** [comment]
@mosababubakr Yes I already did that, but unfortunately I cannot add reviewers to PRs. But please be patient, Cordova maintainers do their work in free time. Dave has done a lot for this plugin recently.

**Comment by mosababubakr** [comment]
@NiklasMerz you are correct, and we are appreciate that, thanks for both.

Regards

**Comment by mosababubakr** [comment]
@NiklasMerz do you test this solution in landscape mode? (Does this solution also fix notch in landscape mode)?

**Comment by mosababubakr** [comment]
@NiklasMerz yes, you are correct, and we are appreciate that, thanks for you both.

Do you test this solution in landscape mode? (Does this solution also fix notch in landscape mode)?

Regards

**Comment by NiklasMerz** [comment]
I did a quick test again with the [test app](https://github.com/dpa99c/cordova-plugin-inappbrowser-test ) and it looks good to me. Please test this yourself with your app. You may need to adjust your HTML and CSS to make it look good. There should be plenty Google results for that.

**Comment by mosababubakr** [comment]
Removing the status bar is not a good idea, not solve the issue, and not the optimal solution.

Let's take SAFARI BROWSER APP as a good example here.

Please try to  to open any website, for e.g.: githubb.com, in any iPhone device, you will see that:

---PORTRAIT MODE---
The status bar appear in correct height with safe area in the top.

---LANDSCAPE MODE---
The status bar is hidden.

Taking into account that safe margin area (the left and the right) was set for iphone x family devices or greater in landscape mode.

*** SAFARI BROWSER APP vs. INAPPBROWSER PLUGIN ***

---PORTRAIT MODE---

![1111](https://user-images.githubusercontent.com/47486787/67144036-c796ee00-f27a-11e9-8de1-9d5bfc556a79.jpg)

--------------------
--- LANDSCAPE MODE---

![222](https://user-images.githubusercontent.com/47486787/67144043-f0b77e80-f27a-11e9-92da-52edbc877db0.jpg)

**Comment by mosababubakr** [comment]
There are 2 problems here, i will explain them well:

First: inappbrowser show an ugly layout from the upper side that merged with statusbar, this should be set to the correct height safe margin from the top, not to modify the css settings of the website, because inappbrowser main function is to open external links, and we can't control external links from 3rd parties.

Second: inappbrowser should set safe area in landscape mode, because iphone x family and greater devices have the camera in the (left or the right) of the screen, and because of that we can't control css files of the external websites, so, inappbrowser should set the safe area margins.
(this may need another PR to fix this issue).

Let's take SAFARI BROWSER APP as a good example here to specify the issue regarding inappbrowser plugin.
Please try to  to open any website, for e.g.: githubb.com, in any iPhone device, you will see that:

---PORTRAIT MODE---
The status bar appear in correct height with safe area in the top.

---LANDSCAPE MODE---
The status bar is hidden.

Taking into account that safe margin area (the left and the right) was set for iphone x family devices or greater in landscape mode.

*** SAFARI BROWSER APP vs. INAPPBROWSER PLUGIN ***

---PORTRAIT MODE---

![1111](https://user-images.githubusercontent.com/47486787/67144036-c796ee00-f27a-11e9-8de1-9d5bfc556a79.jpg)

--------------------
--- LANDSCAPE MODE---

![222](https://user-images.githubusercontent.com/47486787/67144043-f0b77e80-f27a-11e9-92da-52edbc877db0.jpg)

**Comment by sardapv** [comment]
This doesn't fix this is in landscape mode. even though I have viewport-fit=cover with padding of safe-area-inset.. it doesn't work when u change orientation of inappbrowser. however if opened in landscaped mode works fine then rotate to portrait..doesn't! and vice versa. I am not sure how to set height and width of statusBarFrame detecting orientation change..there is transparent area (where OP has grey area) my notification bar of webpage sticks below notch but page content is displayed through transaparent area while scrolling! really bad User experience :(

**Comment by kartiksolanki** [comment]
@mtlehtin : can you please elaborate on how did you configured option for background color and UIStatusBarStyleLightContent or UIStatusBarStyleDefault? i was able to fix the height issue with what you have suggested but i am not able to change the gray background color?

**Comment by mtlehtin** [comment]
Hi, sorry not for answering for these. I've not been involved in the project anymore.

But as for the background-color these are the code changes that i made in our internal fork:

CDVInAppBrowserOptions.h: Expose the options:
```
@property (nonatomic, assign) BOOL statusbarlightcontent;
@property (nonatomic, copy) NSString* backgroundcolor;
```

CDVWKInAppBrowser.m:
```
- (void)createViews
{
  // ...

  // Set background color if user sets it in options
  if (_browserOptions.backgroundcolor != nil) {
    self.view.backgroundColor = [self colorFromHexString:_browserOptions.backgroundcolor];
  } else {
    self.view.backgroundColor = [UIColor whiteColor];
  }

  [self.view addSubview:self.toolbar];
  [self.view addSubview:self.addressLabel];
  [self.view addSubview:self.spinner];
}

- (UIStatusBarStyle)preferredStatusBarStyle
{
    return _browserOptions.statusbarlightcontent ? UIStatusBarStyleLightContent : UIStatusBarStyleDefault;
}
```

Usage in the app:
```
window.cordova.InAppBrowser.open(url, target, 'backgroundcolor=#000000,statusbarlightcontent=yes|no')
```

**Comment by kartiksolanki** [comment]
@mtlehtin : Thanks for the reply. I did those changes and well the status bar does appear but what happens is I have my toolbar on bottom, and the status bar with set color appears in the bottom on top of bottom toolbar instead of status bar appearing on top at its default place. Any idea what is wrong?
